File: smartportfolio/simulation.py
"""Simulation module for SmartPortfolio CLI."""

from dataclasses import dataclass
from typing import Tuple

from .dashboard import Portfolio, Activity, display_dashboard

@dataclass
class Trade:
    ticker: str
    amount: float
    action: str  # "buy" or "sell"


def simulate_trade(portfolio: Portfolio, trade: Trade) -> Portfolio:
    """Return a new portfolio after applying the simulated trade."""
    new_portfolio = Portfolio(
        total_value=portfolio.total_value,
        cash_balance=portfolio.cash_balance,
        activities=list(portfolio.activities),
    )
    if trade.action == "buy":
        new_portfolio.cash_balance -= trade.amount
        new_portfolio.total_value += trade.amount
    else:
        new_portfolio.cash_balance += trade.amount
        new_portfolio.total_value -= trade.amount
    new_portfolio.activities.append(
        Activity(description=f"Simulated {trade.action} {trade.ticker}", amount=trade.amount)
    )
    return new_portfolio


def run_simulation(portfolio: Portfolio) -> None:
    """Interactively simulate a trade."""
    print("\n===== Simulation Mode =====")
    ticker = input("Ticker symbol: ")
    action = input("Action (buy/sell): ").lower()
    amount = float(input("Amount in USD: "))
    trade = Trade(ticker=ticker, action=action, amount=amount)
    new_portfolio = simulate_trade(portfolio, trade)
    display_dashboard(new_portfolio)

File: smartportfolio/simulation.js
const { Portfolio, Activity, displayDashboard } = require('./dashboard');

class Trade {
  constructor(ticker, amount, action) {
    this.ticker = ticker;
    this.amount = amount;
    this.action = action;
  }
}

function simulateTrade(portfolio, trade) {
  const newPortfolio = new Portfolio(
    portfolio.totalValue,
    portfolio.cashBalance,
    [...portfolio.activities]
  );
  if (trade.action === 'buy') {
    newPortfolio.cashBalance -= trade.amount;
    newPortfolio.totalValue += trade.amount;
  } else {
    newPortfolio.cashBalance += trade.amount;
    newPortfolio.totalValue -= trade.amount;
  }
  newPortfolio.activities.push(
    new Activity(`Simulated ${trade.action} ${trade.ticker}`, trade.amount)
  );
  return newPortfolio;
}

async function runSimulation(portfolio, ask) {
  console.log('\n===== Simulation Mode =====');
  const ticker = await ask('Ticker symbol: ');
  const action = (await ask('Action (buy/sell): ')).toLowerCase();
  const amount = parseFloat(await ask('Amount in USD: '));
  const trade = new Trade(ticker, amount, action);
  const newPortfolio = simulateTrade(portfolio, trade);
  displayDashboard(newPortfolio);
  return newPortfolio;
}

module.exports = { Trade, simulateTrade, runSimulation };
